### ISSUE TITLE
Rack-aware base-view pairing

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17147]
The function `get_view_natural_endpoint()` (`db/view/view.cc`) pairs between base replicas and view replicas: If a base token Tb has three replicas [A,B,C] and the corresponding view token for this write has three replicas [X,Y,Z], we pair these in order: A sends a view update to X, B to Y and C to Z.

The function is DC-aware - the above lists of replicas are always inside one DC, we never send view updates across DCs.
However, it seems we missed the fact that this function is **not rack-aware**: If we have three racks (called "availability zones" in AWS) in the above example, we don't want to pair A to X if they are not in the same rack (cross-rack communication is more expensive). Rather, when we have three racks and RF=3, we know that the NTS replication strategy ensures that on both base and view replicas list we have one replica in each rack - and can pair them accordingly, each pair inside a rack.

For the case described above, i.e., RF equals the number of racks, the implementation can be straightforward, in one of two ways:
1. Since the list of replicas is ordered, we can ask the NTS replication strategy to ensure that the list of replicas is always sorted by rack: The first on the list is on rack 1, the second is on rack 2, the third is on rack 3. Since this will be true for both base and view replica lists, the correct pairing is guaranteed without any code changes to `get_view_natural_endpoint()`.
2. If the list of replicas isn't sorted by rack (the case today), we can modify `get_view_natural_endpoint()` to pair inside the rack, if possible. Same logic that we had for "self-pairing".

Everything gets more complex in the less common - but legal - situation where the number of racks and RF is **not** the same. We can have the following problems:
1. We can have more racks than RF. In this case, a tablet replica may migrate from one node on one rack, to another node on a different rack that doesn't yet have a replica. This can break the pairing (previously we had a base and view on the same rack, after the migration we no longer do). This is exactly the same problem that led us to abandon self-pairing in 300e549267b2f51d0afc25b11707ca1d23b441a9. I don't know how to solve this problem.
2. We can have fewer racks than RF. In this case, NTS is forced to allow more than one replica on the same rack, and again means that replicas can also move between racks and cause the same problem described above.

It appears that we always had this rack-unware pairing problem (which is a **performance** problem, not a correctness problem), even with vnodes. We probably can't safely fix it in retrospect in vnodes - it will change the base-view pairing which can lead to ghost rows in unrepaired databases - but we can definitely fix it now for tablet-based replication - this feature is new and doesn't need to work on existing tables.

**Comment by nyh** [comment]
CC @avikivity @dorlaor

**Comment by mykaul** [comment]
I'd focus on the rack = RF (and both = 3) which is a very common scenario, and ignore the rest (read: leave as it is today).

**Comment by nyh** [comment]
> I'd focus on the rack = RF (and both = 3) which is a very common scenario, and ignore the rest (read: leave as it is today). If it is simple enough, I think it is worth fixing for vnodes as well - due to both cost and performance benefits it can bring )

@mykaul good idea. If I add a special case for number of racks equal RF, and if we assume that the number of racks can't be changed changed (I'm not sure if that's true?), we can have a special if() just for this case. This should be pretty easy to code - and slightly harder to test.

I just want to confirm (maybe @tgrabiec can answer): If RF is equal to the number of racks, are we sure that tablets will **never** be migrated between racks - the tablet migrator will avoid putting two replicas of the same tablet on the same rack?

**Comment by bhalevy** [comment]
> There is one problem with solution 1: Whereas in the past the replica order was random, and doing something (like repair on Alternator TTL) on "the primary replica" resulted in good load balancing, with solution 1 the primary replica will always be the one on rack 1, so rather than distributing the load to all nodes, algorithms like primary-range repair and Alternator TTL will suddenly place more load on rack 1.

I think we need to consider that maybe sort on the hash of the rack name and another, de-clustering factor, like the view table id, so different tables and views would have different orders and balance the load on the primary replica more evenly.

Alternatively, would it be feasible to maintain the pairing relationship in system.tablets itself so that when the tablet load balancer migrates a tablet of the base table it would also correspondingly migrate other tablets that are paired with it.
I think this could come handy also for CDC.

**Comment by nyh** [comment]
> I think we need to consider that maybe sort on the hash of the rack name and another, de-clustering factor, like the view table id, so different tables and views would have different orders and balance the load on the primary replica more evenly.

‎Base and views of the same table will need to use the same offset, so this idea will only be effective if there are many different tables in the database. I agree that if there are, it will be effective.

> Alternatively, would it be feasible to maintain the pairing relationship in system.tablets itself so that when the tablet load balancer migrates a tablet of the base table it would also correspondingly migrate other tablets that are paired with it. I think this could come handy also for CDC.

I don't think this idea can work in the general case (where view's partition key is different from the base's) - the point is that for a write to one partition in the base table, theoretically you might need to write to _any_ partition in the view, in any view tablet. There is no pairing between **one** base tablet to **one** view tablet.

**Comment by avikivity** [comment]
It's enough to have the initial placement maintain some sorted order. Migrations, splits and merges will maintain the order. So if we have nr_racks == replication_factor, we'll continue to maintain proximity,

**Comment by nyh** [comment]
> It's enough to have the initial placement maintain some sorted order. 

It's not enough to have "some" sorted order, you need this order to be rack order, because the pairing will choose the first base replica with the first view replica, and if we want them to be in the same rack, the sorting needs to be according to racks. Then we have the problem I mentioned above that the "primary replica" of each tablet is always on rack 1, which breaks code (like repair and Alternator TTL) which today assume the primary replicas are uniformly distributed. I think I have two options now:

1. Assume that both repair and Alternator TTL will stop using the "primary replica" to decide where to do which work. Since both of these still need work on tablets, we can delay this work.
2. Not use option 1 above (sorting the replicas by racks) and use option 2 instead (modification to `get_view_natural_endpoint()`). I mentioned above there is a problem with option 2, maybe we have a "workaround" for that, I'll write a separate comment about that

> Migrations, splits and merges will maintain the order. So if we have nr_racks == replication_factor, we'll continue to maintain proximity,

Migrations and splits are file. I just realized we may have a problem with merges:
With merges, if we have replicas A,B,C for one tablet and X,Y,Z for neighboring tablet, we need to first move them to the same three nodes, before we can do the merge. For maintaining the order, we **must** move A and X to the same node, B and Y to the same node, and C and Z. We cannot allow a different order just because it's more efficient - e.g., if A and Y are already on the same node, it's not good enough! I'll open a separate issue about that.

**Comment by nyh** [comment]
> Unfortunately, after additional discussion with @bhalevy and @avikivity the conclusion is that my statement above "_If we assume that the number of racks can't be changed changed_" turns out to be a bad idea - we can't assume that, and can't base our optimization on this assumption holding.
> 
> This means that solution 2 mentioned above (a modification to `get_view_natural_endpoint()` similar in spirit to how self-pairing works) shares the same problem that self-pairing had: After adding a rack or changing RF, the pairing will break (result in incorrect pairing) - just like self-pairing broke if we had N=RF and then added a node.

I had a discussion with @mykaul who suggested that maybe we can use solution 2 (a modification to `get_view_natural_endpoint()`) and **ignore** the problem of what happens in the (very unlikely!) case of RF or nracks being changed later. Basically, the idea is that we already have a bunch of known **consistency problems** with materialized views, which we'll need some sort of repair-we-don't-have-yet to fix, or a complete view rebuild. Then, the case of "changing RF or nracks when RF=nracks" will just be one of these cases - after you do this change (which we should emphasize is very unlikely!), you **may** have a base-view inconsistency so will need to find a way to repair/rebuild it, as usual.

**Comment by avikivity** [comment]
> > It's enough to have the initial placement maintain some sorted order.
> 
> It's not enough to have "some" sorted order, you need this order to be rack order, because the pairing will choose the first base replica with the first view replica, and if we want them to be in the same rack, the sorting needs to be according to racks. 

Yes, I meant that how racks are sorted is immaterial, so long as it is stable over time.

> Then we have the problem I mentioned above that the "primary replica" of each tablet is always on rack 1, which breaks code (like repair and Alternator TTL) which today assume the primary replicas are uniformly distributed. I think I have two options now:
> 
> 1. Assume that both repair and Alternator TTL will stop using the "primary replica" to decide where to do which work. Since both of these still need work on tablets, we can delay this work.
> 2. Not use option 1 above (sorting the replicas by racks) and use option 2 instead (modification to `get_view_natural_endpoint()`). I mentioned above there is a problem with option 2, maybe we have a "workaround" for that, I'll write a separate comment about that

3. Change the primary endpoint algorithm for tablets. For example, the primary endpoint's index could be hash(end token) % (sum of replication factors). This changes when a tablet is split, but it also changes when vnodes are added, so it's not a behavioral change.

> 
> > Migrations, splits and merges will maintain the order. So if we have nr_racks == replication_factor, we'll continue to maintain proximity,
> 
> Migrations and splits are file. I just realized we may have a problem with merges: With merges, if we have replicas A,B,C for one tablet and X,Y,Z for neighboring tablet, we need to first move them to the same three nodes, before we can do the merge. For maintaining the order, we **must** move A and X to the same node, B and Y to the same node, and C and Z. We cannot allow a different order just because it's more efficient - e.g., if A and Y are already on the same node, it's not good enough! I'll open a separate issue about that.

If A,X started on the same rack, they will remain on the same rack (if RF=nr_racks).

**Comment by nyh** [comment]
> If A,X started on the same rack, they will remain on the same rack (if RF=nr_racks).

The potential problem I mentioned with merges is unrelated to racks. I opened https://github.com/scylladb/scylladb/issues/17265.

**Comment by mykaul** [comment]
@nyh - I doubt that's going to make it to 6.0 - unless it's done in the coming weeks?

**Comment by bhalevy** [comment]
https://github.com/scylladb/scylladb/pull/17846 changes tablet allocation order to make all tablets (on all tables), use the canonical rack order (by rack name).  It just shuffles the node order within racks.

**Comment by tgrabiec** [comment]
> #17846 changes tablet allocation order to make all tablets (on all tables), use the canonical rack order (by rack name). It just shuffles the node order within racks.

I think the order may be broken, e.g. when load balancer finds a replacement on decommission for replicas[0], it may chose any rack which is not already occupied by replicas[1:-1]. If there are overloaded racks (more than 1 replica of a given tablet), it may chose target node from any rack because that will increase rack overload.

**Comment by tgrabiec** [comment]
Btw, we should be documenting expected invariants in one place, maybe https://github.com/scylladb/scylladb/blob/master/docs/dev/topology-over-raft.md?plain=1#L230.

**Comment by mykaul** [comment]
I wonder if we should split this work item to tablets and non-tablets use cases.

Anyway, moving to 6.1.

**Comment by bhalevy** [comment]
> I wonder if we should split this work item to tablets and non-tablets use cases.
> 
> Anyway, moving to 6.1.

I'm not sure it's worth addressing without tablets.

**Comment by bhalevy** [comment]
> > #17846 changes tablet allocation order to make all tablets (on all tables), use the canonical rack order (by rack name). It just shuffles the node order within racks.
> 
> I think the order may be broken, e.g. when load balancer finds a replacement on decommission for replicas[0], it may chose any rack which is not already occupied by replicas[1:-1]. If there are overloaded racks (more than 1 replica of a given tablet), it may chose target node from any rack because that will increase rack overload.

Most importantly, we should ensure the rack order when RF = Number of racks, since this should the most common case in a multi DC/RACK setup.

**Comment by mykaul** [comment]
> > I wonder if we should split this work item to tablets and non-tablets use cases.
> > Anyway, moving to 6.1.
> 
> I'm not sure it's worth addressing without tablets.

If we go with the 'easy' case of RF=3 and 3 racks, which I think is a very very common scenario, then it's an important win in terms of cost and perhaps performance.
Of course I'd expect a follow-up for tablets. If it's in the same code path and easier to accomplish together - sure.

**Comment by bhalevy** [comment]
> There is one problem with solution 1: Whereas in the past the replica order was random, and doing something (like repair on Alternator TTL) on "the primary replica" resulted in good load balancing, with solution 1 the primary replica will always be the one on rack 1, so rather than distributing the load to all nodes, algorithms like primary-range repair and Alternator TTL will suddenly place more load on rack 1.

This should have been solved in f133ae945afdddc9d95a6f7b91a30db42337f4bd

**Comment by bhalevy** [comment]
> > [#17846](https://github.com/scylladb/scylladb/pull/17846) changes tablet allocation order to make all tablets (on all tables), use the canonical rack order (by rack name). It just shuffles the node order within racks.
> 
> I think the order may be broken, e.g. when load balancer finds a replacement on decommission for replicas[0], it may chose any rack which is not already occupied by replicas[1:-1]. If there are overloaded racks (more than 1 replica of a given tablet), it may chose target node from any rack because that will increase rack overload.

@tgrabiec doesn't https://github.com/scylladb/scylladb/issues/19475 disprove that theory?

**Comment by tgrabiec** [comment]
> > > [#17846](https://github.com/scylladb/scylladb/pull/17846) changes tablet allocation order to make all tablets (on all tables), use the canonical rack order (by rack name). It just shuffles the node order within racks.
> > 
> > 
> > I think the order may be broken, e.g. when load balancer finds a replacement on decommission for replicas[0], it may chose any rack which is not already occupied by replicas[1:-1]. If there are overloaded racks (more than 1 replica of a given tablet), it may chose target node from any rack because that will increase rack overload.
> 
> [@tgrabiec](https://github.com/tgrabiec) doesn't [#19475](https://github.com/scylladb/scylladb/issues/19475) disprove that theory?

Can you explain how? I guess in #19475 we have a case where racks were not overloaded before decommission, and that's why decommission cannot find a new replica. In my comment you quoted I'm talking about the case when racks were already overloaded before decommission, so movement to a rack which already has 1 replica is allowed.
